### ISSUE TITLE
CSVFIleReader set all fields as required in Avro

### ISSUE THREAD

**Issue by cjtravis** [mmolimar/kafka-connect-fs#83]
Currently testing the connector using CSVFileReader to read from S3.  The connector works fine to read from the CSV and load into Kafka when all fields in the CSV are populated.

Connector configuration looks like this:
```json
{
  "name" : "kafka-connect-s3-demo",
  "config" : {
    "connector.class" : "com.github.mmolimar.kafka.connect.fs.FsSourceConnector",
    "file_reader.class" : "com.github.mmolimar.kafka.connect.fs.file.reader.CsvFileReader",
    "file_reader.delimited.settings.format.delimiter" : ",",
    "file_reader.delimited.settings.header" : true,
    "file_reader.delimited.settings.allow_nulls" : true,
    "fs.uris" : "s3a://bucket/prefix/",
    "name" : "kafka-connect-s3-demo",
    "policy.class" : "com.github.mmolimar.kafka.connect.fs.policy.SleepyPolicy",
    "policy.fs.fs.s3a.access.key" : "KEY",
    "policy.fs.fs.s3a.secret.key" : "SECRET",
    "policy.regexp" : "^test_.*000",
    "policy.sleepy.sleep" : "60000",
    "topic" : "s3.leads",
    "key.converter" : "org.apache.kafka.connect.storage.StringConverter",
    "value.converter" : "io.confluent.connect.avro.AvroConverter",
    "value.converter.schema.registry.url" : "http://sr-url:8081"
  }
}
```

However, the moment I try sending through a file that contains a null value for one of the fields, the connector registers a new version in schema registry requiring all fields and yielding this error in the SR log:

`bash
kafka-connect       | [2021-02-25 13:51:45,618] INFO FsSourceTask Processing records for file [path = s3a://bucket/prefix/test_000.csv, length = 496, blocks = [[offset = 0, length = 496, corrupt = false]]]... (com.github.mmolimar.kafka.connect.fs.FsSourceTask)
kafka-connect       | [2021-02-25 13:51:45,618] WARN FsSourceTask Error reading file [s3a://bucket/prefix/test_000.csv]: Invalid value: null used for required field: "contact2firstnm", schema type: STRING. Keep going... (com.github.mmolimar.kafka.connect.fs.FsSourceTask)
kafka-connect       | org.apache.kafka.connect.errors.DataException: Invalid value: null used for required field: "contact2firstnm", schema type: STRING
kafka-connect       | 	at org.apache.kafka.connect.data.ConnectSchema.validateValue(ConnectSchema.java:220)
kafka-connect       | 	at org.apache.kafka.connect.data.Struct.put(Struct.java:216)
kafka-connect       | 	at org.apache.kafka.connect.data.Struct.put(Struct.java:203)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.file.reader.UnivocityFileReader$UnivocityToStruct.lambda$apply$1(UnivocityFileReader.java:268)
kafka-connect       | 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfInt.accept(ForEachOps.java:204)
kafka-connect       | 	at java.base/java.util.stream.IntPipeline$9$1.accept(IntPipeline.java:369)
kafka-connect       | 	at java.base/java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:104)
kafka-connect       | 	at java.base/java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:699)
kafka-connect       | 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
kafka-connect       | 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
kafka-connect       | 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
kafka-connect       | 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfInt.evaluateSequential(ForEachOps.java:188)
kafka-connect       | 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
kafka-connect       | 	at java.base/java.util.stream.IntPipeline.forEach(IntPipeline.java:439)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.file.reader.UnivocityFileReader$UnivocityToStruct.apply(UnivocityFileReader.java:265)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.file.reader.UnivocityFileReader$UnivocityToStruct.apply(UnivocityFileReader.java:258)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.file.reader.AbstractFileReader.next(AbstractFileReader.java:94)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.file.reader.AbstractFileReader.next(AbstractFileReader.java:18)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.FsSourceTask.lambda$poll$1(FsSourceTask.java:95)
kafka-connect       | 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
kafka-connect       | 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
kafka-connect       | 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
kafka-connect       | 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
kafka-connect       | 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
kafka-connect       | 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
kafka-connect       | 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
kafka-connect       | 	at com.github.mmolimar.kafka.connect.fs.FsSourceTask.poll(FsSourceTask.java:109)
kafka-connect       | 	at org.apache.kafka.connect.runtime.WorkerSourceTask.poll(WorkerSourceTask.java:289)
kafka-connect       | 	at org.apache.kafka.connect.runtime.WorkerSourceTask.execute(WorkerSourceTask.java:256)
kafka-connect       | 	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
kafka-connect       | 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
kafka-connect       | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
kafka-connect       | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
kafka-connect       | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
kafka-connect       | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
kafka-connect       | 	at java.base/java.lang.Thread.run(Thread.java:834)
`

Here is a snippet reflecting the change in the Avro schema that the connector registers:
Avro schema before encountering a row with a `null` value:
```json
{
   "name":"contact1firstnm",
   "type":[
      "null",
      "string"
   ],
   "default":null
}
```
Avro schema after encountering row with a `null` value:
```json

{
   "name":"contact1firstnm",
   "type":"string"
}
```

My expectation was that with `file_reader.delimited.settings.allow_nulls` set to `true` that would permit nulls being passed in any of the fields within the CSV.  Either I misunderstood or that's not happening as designed. 

In short, I'd like to be able to load a CSV from S3 into Kafka where the fields may be null from row to row.

**Comment by mmolimar** [comment]
Hi!
As you said, allowing nulls, the schema would be the same as in your first example because the fields will be optional based on the properties ``file_reader.delimited.settings.allow_nulls`` (true) and/or ``file_reader.delimited.settings.data_type_mapping_error`` (false). So it should work.

Have you tried with ``file_reader.delimited.settings.null_value`` and/or ``file_reader.delimited.settings.empty_value`` properties?

**Comment by cjtravis** [comment]
@mmolimar Thanks for the quick response.  Good news it appears.  Explicitly setting:
```json
"file_reader.delimited.settings.null_value": "null",
"file_reader.delimited.settings.empty_value": "null",
```
in the properties now permits those values to come across.  I didn't catch in the documentation where those needed to be explicitly set since they already defaulted to `null`.  Either way, that seems to have cleared up my concern there.  One quick follow up: is there a way to override or set a custom `namespace` in the Avro definition.  By default I observe it defaults to `io.confluent.connect.avro`.

Thanks again for the help.

**Comment by mmolimar** [comment]
Great!

Not in the case of this file reader. I think what you need is to apply a SMT like [this](https://docs.confluent.io/platform/current/connect/transforms/setschemametadata.html).

**Comment by cjtravis** [comment]
Perfect, I can kick the tires on that SMT.  Thanks so much for the help today!
